autonomous/score_and_balance.py

```diff
@@ -0,0 +1,210 @@
+from magicbot.state_machine import AutonomousStateMachine, state
+from wpimath.geometry import Rotation2d, Translation2d, Pose2d
+from dataclasses import dataclass
+from components.arm import Arm
+from components.gripper import Gripper
+
+from controllers.movement import Movement
+from controllers.recover import RecoverController
+from controllers.score_game_piece import ScoreGamePieceController
+
+from utilities.game import (
+    Node,
+    Rows,
+    field_flip_rotation2d,
+    field_flip_translation2d,
+    field_flip_pose2d,
+    get_score_location,
+    is_red,
+)
+
+
+@dataclass
+class PickupAction:
+    piece_idx: int
+    approach_direction: Rotation2d
+    intermediate_waypoints: tuple[Translation2d, ...]
+
+    def with_correct_flipped(self) -> "PickupAction":
+        if is_red():
+            waypoints = tuple(
+                field_flip_translation2d(x) for x in self.intermediate_waypoints
+            )
+            return PickupAction(
+                self.piece_idx,
+                field_flip_rotation2d(self.approach_direction),
+                waypoints,
+            )
+        return self
+
+
+@dataclass
+class ScoreAction:
+    node: Node
+    intermediate_waypoints: tuple[Translation2d, ...]
+
+    def with_correct_flipped(self) -> "ScoreAction":
+        if is_red():
+            waypoints = tuple(
+                field_flip_translation2d(x) for x in self.intermediate_waypoints
+            )
+            return ScoreAction(self.node, waypoints)
+        return self
+
+
+@dataclass
+class FieldPose:
+    pose: Pose2d
+    approach_angle: Rotation2d
+    intermediate_waypoints: tuple[Translation2d, ...]
+
+    def with_correct_fipped(self) -> "FieldPose":
```

Comment:
what's a fipped?

Reply:
Turns out I couldn't spell, should have been "flipped"

Reply:
may I suggest enabling spell check in your editor?

Reply:
Yea okay